src/vtk_calculator.py

```diff
@@ -1,53 +1,106 @@
-    data = mesh_io.read_dataset(args.in_meshname)
-    logging.info("Read in {} points.".format(data.GetNumberOfPoints()))
+
+    out_meshname = args.out_meshname
+    if args.out_meshname is None:
+        out_meshname = args.in_meshname
+
+    reader = vtk.vtkGenericDataObjectReader()
+    reader.SetFileName(args.in_meshname)
+    reader.Update()
+    vtk_dataset = reader.GetOutput()
+    logging.info("Read in {} points.".format(vtk_dataset.GetNumberOfPoints()))
 
     calc = vtk.vtkArrayCalculator()
-    calc.SetInputData(data)
+    calc.SetInputData(vtk_dataset)
     calc.AddCoordinateScalarVariable("coordsX", 0)
     calc.AddCoordinateScalarVariable("coordsY", 1)
     calc.AddCoordinateScalarVariable("coordsZ", 2)
     calc.SetFunction(args.function)
-    calc.SetResultArrayName("scalars")
-    calc.Update()
+    calc.SetResultArrayName(args.tag)
+    calc.Update()    
     logging.info("Evaluated '{}' on the mesh.".format(args.function))
+    
+    if args.diff:
+        assert args.intag is not None, "Difference mode needs input array tag"
 
-    writer = vtk.vtkUnstructuredGridWriter()
-    writer.SetInputData(calc.GetOutput())
-    writer.SetFileName(args.out_meshname)
-    writer.Write()
-    logging.info("Written output to {}.".format(args.out_meshname))
+        org_val = []
+        calc_val = []
+        orgArr = calc.GetOutput().GetPointData().GetAbstractArray(args.intag)
+        calcArr = calc.GetOutput().GetPointData().GetAbstractArray(args.tag)
+        num_points = vtk_dataset.GetNumberOfPoints()
+        for i in range(num_points):
+            org_val.append(orgArr.GetTuple1(i))
+            calc_val.append(calcArr.GetTuple1(i))
+        
+        org_val = np.array(org_val)
+        calc_val = np.array(calc_val)
+        difference = np.abs(org_val - calc_val)
+        cnt, min, max = num_points, np.nanmin(difference), np.nanmax(difference)
+        p99, p95, p90, median = np.percentile(difference, [99, 95, 90, 50])
 
+        logging.info("Vertex count {}".format(cnt))
+        logging.info("Maximum error per vertex {}".format(max))
+        logging.info("Minimum error per vertex {}".format(min))
+        logging.info("Median error per vertex {}".format(median))
+        logging.info("99th percentile of error per vertex {}".format(p99))
+        logging.info("95th percentile of error per vertex {}".format(p95))
+        logging.info("90th percentile of error per vertex {}".format(p90))
 
+        if args.stats:
+            base = os.path.splitext(out_meshname)[0]
+            json.dump({
+                "count": cnt,
+                "min": min,
+                "max": max,
+                "median": median,
+                "99th percentile": p99,
+                "95th percentile": p95,
+                "90th percentile": p90
+            }, open(base + ".stats.json", "w"))
 
+    writer = vtk.vtkUnstructuredGridWriter()
+    writer.SetInputData(calc.GetOutput())
```

Comment:
Does this preserve all other PointData?

Reply:
Calculator object appends calculated data to a given dataset, hence all previous variables are preserved. If the tag is already used it overrides it.